### PR TITLE
Add piecewise script from VPREC paper

### PR DESCRIPTION
**Pull request by yohanchatelain**
This pull request introduces a new Python tool, `vfc_piecewise`, for searching piecewise virtual precision schedules and integrates it into the Verificarlo Python ecosystem. It also improves documentation, enhances the build and install process to support the `uv` Python package manager, and updates several test and ignore files for better developer experience.

**New piecewise precision search tool:**

* Added `src/tools/optimize/piecewise.py`, a new command-line tool and Python API (`vfc_piecewise`) for searching minimal per-step virtual precision schedules in iterative programs, supporting multiple strategies and animated visualizations.
* Registered `vfc_piecewise` as a Verificarlo CLI script and included the new tool in the Python build system (`pyproject.toml`). [[1]](diffhunk://#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R39) [[2]](diffhunk://#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R51-R64)

**Documentation improvements:**

* Documented the new `vfc_piecewise` tool in the user guide and postprocessing documentation, including usage instructions, strategies, and example pipeline. [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R159) [[2]](diffhunk://#diff-0a85967468fc45cb68d20b7e8d9665da0a891fe8b4c3ae20b498d2feeffabf5dR40-R101) [[3]](diffhunk://#diff-afa7cdf8b6f46dabb83bd9c1b012b0c83b05ac0c4161b44c2a59efb49d4e548eL71-R90)

**Python packaging and dependencies:**

* Added `matplotlib` as a required dependency to support `vfc_piecewise` animation features.

**Build and install enhancements:**

* Added support for the `uv` package manager in the build system (`configure.ac`, `Makefile.am`), and updated the install documentation to recommend `uv` for Python installs. The install process now uses `uv` if available, falling back to `pip` otherwise. [[1]](diffhunk://#diff-49473dca262eeab3b4a43002adb08b4db31020d190caaad1594b47f1d5daa810L240-R248) [[2]](diffhunk://#diff-0462e381b2fb3286568215681c8983490a37ac9ae0f0c5ee304df7fa6426d4afL35-R43) [[3]](diffhunk://#diff-64753026d17294e1abe3e21c17bc4ad63b495c197263a70146b170e087840c2aR39-R53) [[4]](diffhunk://#diff-2f754321d62f08ba8392b9b168b83e24ea2852bb5d815d63e767f6c3d23c6ac5R1-R6) [[5]](diffhunk://#diff-afa7cdf8b6f46dabb83bd9c1b012b0c83b05ac0c4161b44c2a59efb49d4e548eL71-R90)

**Testing and developer workflow:**

* Improved `.dockerignore` and `.gitignore` files for Python and test artifacts, and updated a test script to handle timeouts and debuginfod edge cases more robustly. [[1]](diffhunk://#diff-2f754321d62f08ba8392b9b168b83e24ea2852bb5d815d63e767f6c3d23c6ac5R1-R6) [[2]](diffhunk://#diff-d3df2e75cba7371c50176ae21f74c8504b179c83eac66a9a9f0af873d23a6d4aR1-R4) [[3]](diffhunk://#diff-1060f8a392b550d97fa1da7c6a3c00174688d07fe99f59f037384ab2e993bf28R8) [[4]](diffhunk://#diff-cdad5f4717dd194405f51cbf8d64e31e4ece5220aa6a4b7173e23ba92620503eR5-R24) [[5]](diffhunk://#diff-c318ead67220fcf75b54d0ea1c5453f6071b62541e4d47cce54de42e0a5e6e36L182-R183)

These changes collectively add a powerful new tool for precision analysis, streamline Python environment management, and improve developer experience.

### CODE REVIEW

File: .github/workflows/docker-ci.yml [hunk 1]
```diff
@@ -36,6 +36,21 @@ jobs:
       - name: Test
         run: |
           docker run --rm ${{ env.TEST_TAG }} /bin/bash -c "make -C /build/verificarlo installcheck"
+      - name: Verify uv is used for the Python install when available
```
Comment (pablooliveira) [inline .github/workflows/docker-ci.yml:39]: Do we want to have such tests in the docker CI?

The `Test` target runs the verificarlo tests on the built images to ensure they are passing tests. But the new workflow seems to be more of a unit test. I understand it's useful, but also worry it's complicating the docker-ci.yml?

Not sure what to do here. What do you think?

File: doc/01-Install.md [hunk 2]
```diff
@@ -68,7 +68,26 @@ On AArch64 platforms where PRISM backend dependencies may fail to build, use the
...
+   ```bash
+   $ source .venv/bin/activate
+   ```
+
+3. Install Verificarlo and dependencies:
+   ```bash
+   $ uv pip install .
+   ```
+
+### Installing Verificarlo in a standard Python virtual environment (venv)
```
Comment (pablooliveira) [inline doc/01-Install.md]: In the documentation I would remove the section for venv (to keep things short) since we recommend uv and just add a short note saying that Verificarlo also supports virtual environments created using venv.

File: doc/06-Postprocessing.md [hunk 3]
```diff
@@ -37,6 +37,70 @@ and ``AllOpsResults.csv`` for arguments and internal operations.
...
+
+For the full list of options: `vfc_piecewise --help`.
+
+The [`tests/test_newton_vprec/`](https://github.com/verificarlo/verificarlo/tree/master/tests/test_newton_vprec/)
+test demonstrates a complete pipeline: a Newton-Raphson C program instrumented
+with `interflop_call`, a `piecewise_search.py` script that drives all three
+strategies through the `PiecewiseSearchOptimizer` Python API directly (useful
+if you need more control than the CLI provides), and a `plot_results.py`
+script that plots the resulting schedules against the achieved convergence.
+
```
Comment (pablooliveira) [inline doc/06-Postprocessing.md:103]: This is great. I think it's great to have such script in verificarlo `src\tools/`. 

I used a custom script recently for such an exploration, one important parameter was

```python
   parser.add_argument("--max-extra-iterations", type=int, nargs='+', default=[0], help="Maximum number of extra iterations allowed compared to baseline.")
```

The idea is that often if you allow a few extra iterations, you can achieve much lower precision profiles.
From my understanding, in your PR, all the checks are done through the exit code of the tested program. So giving an extra iterations budget is in the hands of the user when he writes his oracle.

But maybe we could add a note here?

File: tests/test_uv_integration/test.sh [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+# Regression test for the `uv` package manager integration
+# (configure.ac UV detection + Makefile.am install-exec-hook).
+#
+# It does not run the full ./configure (LLVM detection, etc. make this
+# slow and environment-dependent); instead it regenerates `configure`
+# from configure.ac with `autoconf` alone and checks that the variables
+# the install-exec-hook depends on are actually wired up. This is what
+# catches issues like a missing AC_SUBST silently breaking `pip install`
+# on Python >= 3.11 without uv installed.
```
Comment (pablooliveira) [inline tests/test_uv_integration/test.sh:12]: Is this test for uv enough? Can we remove the extra test workflow from the docker-ci ?

**Review by pablooliveira — COMMENTED** [review]
Thanks! Looks great, submitted a few comments.

File: doc/06-Postprocessing.md (hunk 3 above)
Comment (yohanchatelain) [reply doc/06-Postprocessing.md:103]: Added a dedicated note in `doc/06-Postprocessing.md` explaining how iteration budgets interact with convergence in iterative solvers, and how users can adjust their program's iteration bound along with the schedule size (`-n`) to let `vfc_piecewise` explore lower precision profiles.

File: tests/test_uv_integration/test.sh (hunk 4 above)
Comment (yohanchatelain) [reply tests/test_uv_integration/test.sh:12]: They test two different layers:

- `tests/test_uv_integration/test.sh` is a fast, static sanity check that runs during `make check` without needing `uv` installed; it checks autoconf variable substitutions and `Makefile.am` branching logic.
- The `docker-ci.yml` step is an end-to-end integration test: it dynamically installs `uv` in the container image, reconfigures, and runs `make install-exec-hook` to verify that `uv pip install --system --break-system-packages` actually succeeds against real PEP 668 environments and isn't broken by issues like host venv leakage.

So keeping the Docker CI step is useful to catch runtime packaging regressions.

File: doc/06-Postprocessing.md (hunk 3 above)
Comment (pablooliveira) [reply doc/06-Postprocessing.md:103]: Looks good. I suppose that when exploring, any extra iterations will stay on the last used virtual precision?

**Review by pablooliveira — APPROVED** [review]
Thanks for taking into account the feedback.

File: doc/06-Postprocessing.md (hunk 3 above)
Comment (yohanchatelain) [reply doc/06-Postprocessing.md:103]: Yes, exactly. And if the user wants vfc_piecewise to actively search and vary precision across those extra iterations too, they can increase -n to cover the full iteration budget.
